### PR TITLE
Network Interface API

### PR DESCRIPTION
**Pull request by isalento**
Before running live tests, make sure you have virtual network called "jclouds-virtual-network-live-test" and default subnet created in Azure portal. (NIC is created using this default subnet)

To run live tests: 
mvn -Dtest=NetworkInterfaceCardApiLiveTest -Dtest.azurecompute-arm.identity="12345678-1a23-1234-1234-123456789123" -Dtest.azurecompute-arm.subscriptionid="12345678-1234-1234-1a23-123456789abc" -Dtest.azurecompute-arm.resourcegroup="armlivetesting" -Dtest.azurecompute-arm.credential="" -Dtest.azurecompute-arm.endpoint="https://management.azure.com/" -Dtest.jclouds.oauth.resource="https://management.azure.com/" -Dtest.oauth.endpoint="https://login.microsoftonline.com/.onmicrosoft.com/oauth2/token" test

### CODE REVIEW

File: azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/IpConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+public abstract class IpConfiguration{
+
+    @AutoValue
+    public abstract static class IpConfigurationProperties{
+
+        public IpConfigurationProperties() {
+        }// For AutoValue only!
+
+        @Nullable
+        public abstract String provisioningState();
```
Comment (ritazh) [inline azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/IpConfiguration.java:34]: For properties that are not added here like `publicIPAddress`, `loadBalancerBackendAddressPools`, `loadBalancerInboundNatRules`, please make sure they are optional. If so, let's add a low-priority card on Trello.

File: azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/NetworkInterfaceCard.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+        public NetworkInterfaceCardProperties() {
+        }// For AutoValue only!
+
+        @Nullable
+        public abstract String provisioningState();
+
+        @Nullable
+        public abstract String resourceGuid();
+
```
Comment (ritazh) [inline azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/NetworkInterfaceCard.java:42]: Do we need `dnsSettings` and `networkSecurityGroup` now?

File: azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/NetworkInterfaceCard.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    } // For AutoValue only!
+
+    @Nullable
+    public abstract String name();
+
+    @Nullable
+    public abstract String id();
+
+    @Nullable
+    public abstract String etag();
```
Comment (ritazh) [inline azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/NetworkInterfaceCard.java:116]: Tags?

File: azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApi.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+   @GET
+   @Fallback(EmptyListOnNotFoundOr404.class)
+   List<NetworkInterfaceCard> listNetworkInterfaceCards();
+
+   @Named("networkinterfacecard:create_or_update")
+   @Path("networkInterfaces/{networkinterfacecardname}")
+   @MapBinder(BindToJsonPayload.class)
+   @PUT
+   @Fallback(NullOnNotFoundOr404.class)
+   NetworkInterfaceCard createOrUpdateNetworkInterfaceCard(@PathParam("networkinterfacecardname") String networkinterfacecardname,
```
Comment (ritazh) [inline azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApi.java:63]: tags should be here as well

File: azurecompute-arm/src/test/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApiLiveTest.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    @AfterClass
+    public void cleanup(){
+        //TODO: Needs to delete the virtual network & subnet if it was created.
+    }
+
+    @Test(groups = "live")
+    public void deleteNetworkInterfaceCardResourceDoesNotExist() {
+
+        final NetworkInterfaceCardApi nicApi = api.getNetworkInterfaceCardApi(subscriptionid, resourcegroup);
```
Comment (ritazh) [inline azurecompute-arm/src/test/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApiLiveTest.java:67]: Returns 204

File: azurecompute-arm/src/test/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApiLiveTest.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        assertTrue(nicList.size() > 0);
+    }
+
+
+    //mvn -Dtest=VirtualNetworkApiLiveTest#deleteVirtualNetwork test
+    @Test(groups = "live", dependsOnMethods = "listNetworkInterfaceCards", alwaysRun = true)
+    public void deleteNetworkInterfaceCard() {
+
+        final NetworkInterfaceCardApi nicApi = api.getNetworkInterfaceCardApi(subscriptionid, resourcegroup);
+
```
Comment (ritazh) [inline azurecompute-arm/src/test/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApiLiveTest.java:142]: Return 202

File: azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/IpConfiguration.java (hunk 1 above)
Comment (isalento) [inline azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/IpConfiguration.java:34]: They are optional when creating a new NIC, but I think we might need Public IP address along the road. Adding ticket to trello.

File: azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/NetworkInterfaceCard.java (hunk 2 above)
Comment (isalento) [inline azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/domain/NetworkInterfaceCard.java:42]: dnsSettings is needed only if we want to add support for custom DNS servers.
networkSecurityGroup should not be needed at this phase. We would need to add support for corresponding API as well: https://msdn.microsoft.com/en-us/library/azure/mt163656.aspx

File: azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApi.java (hunk 4 above)
Comment (isalento) [inline azurecompute-arm/src/main/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApi.java:63]: Adding tag support.

File: azurecompute-arm/src/test/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApiLiveTest.java (hunk 5 above)
Comment (isalento) [inline azurecompute-arm/src/test/java/org/jclouds/azurecomputearm/features/NetworkInterfaceCardApiLiveTest.java:67]: Started using Janne's @ResponseParser(StatusCodeParser.class) for delete operation.

**Comment by ritazh** [comment]
LGTM

**Pull request merged**